### PR TITLE
Make SQLFluff compatible with DBT 1.5

### PR DESCRIPTION
**Pull request by WittierDinosaur**
### Brief summary of the change made
Fixes an issue wherein dbt.flags cannot find PROFILES_DIR.

The issue that this call used to fix seems to have been fixed in a backported change https://github.com/dbt-labs/dbt-core/commit/c952d44ec5c2506995fbad75320acbae49125d3d#diff-b99c74c2d09fa1ea054aa843fa3738bd7c9d39acb9c5b90ca920f0e1fcf38010

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 1]
```diff
@@ -65,8 +64,9 @@ class DbtConfigArgs:
     profiles_dir: Optional[str] = None
     profile: Optional[str] = None
     target: Optional[str] = None
+    threads: int = 1
     single_threaded: bool = False
-    vars: str = ""
+    vars: dict = None
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:69]: Nit: This should be `Dict`, not `dict`. I think support for the latter was only added in Python 3.10.

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 2]
```diff
@@ -65,8 +64,9 @@ class DbtConfigArgs:
     profiles_dir: Optional[str] = None
     profile: Optional[str] = None
     target: Optional[str] = None
+    threads: int = 1
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:67]: What's this about? It seems unrelated.

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 3]
```diff
@@ -97,21 +97,15 @@ def dbt_version(self):  # pragma: no cover
...
         flags.set_from_args(
             DbtConfigArgs(
                 project_dir=self.project_dir,
                 profiles_dir=self.profiles_dir,
                 profile=self._get_profile(),
                 vars=self._get_cli_vars(),
+                threads=self._get_threads(),
             ),
-            user_config,
+            user_config=None,
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:108]: For safety (I think), let's keep it as a positional parameter. Otherwise, we may be passing it as a different parameter after this change.
```suggestion
            None,
```

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 4]
```diff
@@ -248,7 +243,11 @@ def _get_cli_vars(self) -> str:
             (self.templater_selector, self.name, "context")
         )
 
-        return str(cli_vars) if cli_vars else "{}"
+        return cli_vars if cli_vars else dict()
+
+    def _get_threads(self) -> int:
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:248]: What is the `threads` parameter to dbt? I feel uncomfortable reusing the SQLFluff concurrent linting setting for something that looks different (processes vs threads, SQLFluff linting vs ???).

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py (hunk 1 above)
Comment (WittierDinosaur) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:69]:
```suggestion
    vars: Dict = None
```

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 5]
```diff
@@ -65,8 +64,9 @@ class DbtConfigArgs:
     profiles_dir: Optional[str] = None
     profile: Optional[str] = None
     target: Optional[str] = None
+    threads: int = 1
     single_threaded: bool = False
-    vars: str = ""
+    vars: Dict = None
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:69]: Seems like it should be `Optional [Dict]`, since we're setting it to `None`

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py (hunk 4 above)
Comment (barrywhart) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:248]: Seems like the whole function could be removed, perhaps?

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 6]
```diff
@@ -97,21 +97,15 @@ def dbt_version(self):  # pragma: no cover
...
-        # was suggested and described here:
-        # https://github.com/sqlfluff/sqlfluff/issues/2253#issuecomment-1018722979
-        user_config = read_user_config(flags.PROFILES_DIR)
         flags.set_from_args(
             DbtConfigArgs(
                 project_dir=self.project_dir,
                 profiles_dir=self.profiles_dir,
                 profile=self._get_profile(),
                 vars=self._get_cli_vars(),
+                threads=self._get_threads(),
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:106]:
```suggestion
                threads=1,
```

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 7]
```diff
@@ -248,7 +243,7 @@ def _get_cli_vars(self) -> str:
             (self.templater_selector, self.name, "context")
         )
 
-        return str(cli_vars) if cli_vars else "{}"
+        return cli_vars if cli_vars else {}
```
Comment (barrywhart) [inline plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:246]: I'm confused about this change. The function type says it returns `str`, but we removed a `str()` conversion as well as changing the `else` to return a dictionary. Why make this change? Is it related to the dbt issue?

Comment (WittierDinosaur) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:246]: Yeah, it wants this through as a dict now. I've just committed a change to the type hint

**Review by greg-finley — APPROVED** [review]
👏

**Comment by dbeatty10** [comment]
Did this get released yet?

Ran into this today when using sqlfluff for the first time. Downgrading from dbt 1.5.0 to 1.4.x allowed it to work again.

**Comment by davidculley** [comment]
I still get this error message:

```
AttributeError: module 'dbt.flags' has no attribute 'PROFILES_DIR'
```

I am using

- dbt-core 1.5.0
- dbt-snowflake 1.5.0
- sqlfluff 2.1.0

Downgrading to

- dbt-core (1.5.0 -> 1.4.6)
- dbt-snowflake (1.5.0 -> 1.4.2)

only lead to this error instead:

```
yaml/cyaml.py", line 26, in __init__
    CParser.__init__(self, stream)
  ^^^^^^^^^^^^^^^^^
  File "yaml/_yaml.pyx", line 288, in yaml._yaml.CParser.__init__
TypeError: a string or stream input is required
```

Only downgrading to

- dbt-core (1.5.0 -> 1.4.6)
- dbt-snowflake (1.5.0 -> 1.4.2)
- sqlfluff (2.1.0 -> 2.0.7)
- sqlfluff-templater-dbt (2.1.0 -> 2.0.7)

solved the issue.

Therefore, it seems that SQLFluff 2.1.0 is not yet compatible with dbt 1.5.0.

Thanks for your work so far!

**Comment by tlfbrito** [comment]
I still also get this error using BigQuery adapter for dbt

**Comment by elyobo** [comment]
Same as @davidculley, attempting an upgrade to dbt 1.5 and sqlfluff 2.1.0 results in `AttributeError: module 'dbt.flags' has no attribute 'PROFILES_DIR'`.

**Comment by Sunnyinho** [comment]
echoing @elyobo getting same error `AttributeError: module 'dbt.flags' has no attribute 'PROFILES_DIR'`

**Comment by mvaerle** [comment]
I have the same issue, a (temporary) workaround for my machine is adding the following to `.sqlfluff`:
```
[sqlfluff:templater:dbt]
profiles_dir = ~/.dbt
```
This of course depends on where your dbt profiles actually are, for mac os this is the default location. Also note that if you run this on CI you need to override the setting again, I fixed this by having another sqlfluff config file and passing `--profile ` to sqlfluff in CI.

**Comment by JoeryV** [comment]
I get the `AttributeError: module 'dbt.flags' has no attribute 'PROFILES_DIR'` error as well.
Error gets attributed to the file `sqlfluff_templater_dbt/templater.py` [line 196](https://github.com/sqlfluff/sqlfluff/blob/c6ffb83ab7f7d20a86b859bcb84c4dde015ea598/plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py#LL196C33-L196C33)

**Comment by greg-finley** [comment]
For further discussion, see:
https://github.com/sqlfluff/sqlfluff/pull/4886
or
https://github.com/sqlfluff/sqlfluff/issues/4841
